exp/operator/controllers/coreprovider_controller.go

```diff
@@ -19,31 +19,85 @@ package controllers
 import (
 	"context"
 
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	operatorv1 "sigs.k8s.io/cluster-api/exp/operator/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/util/patch"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller"
-
-	operatorv1alpha4 "sigs.k8s.io/cluster-api/exp/operator/api/v1alpha4"
 )
 
 // CoreProviderReconciler reconciles a CoreProvider object.
 type CoreProviderReconciler struct {
 	Client client.Client
 }
 
+func (r *CoreProviderReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&operatorv1.CoreProvider{}).
+		WithOptions(options).
+		Complete(r)
+}
+
 // +kubebuilder:rbac:groups=operator.cluster.x-k8s.io,resources=coreproviders,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operator.cluster.x-k8s.io,resources=coreproviders/status,verbs=get;update;patch
 
-func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	log := ctrl.LoggerFrom(ctx)
-	log.V(4).Info("Reconcile CoreProvider")
+func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
```

Comment:
Wondering if we can avoid 4 reconcilers. What if we have a 

```go
type GenericProviderReconciler struct {
	Object client.Object
	Client client.Client
}

func (r *GenericProviderReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
	return ctrl.NewControllerManagedBy(mgr).
		For(r.Object).
		WithOptions(options).
		Complete(r)
}

func (r *GenericProviderReconciler) Reconcile(ctx context.Context, req reconcile.Request) (_ reconcile.Result, reterr error) {
	typedProvider := r.Object.DeepCopyObject()
	if err := r.Client.Get(ctx, req.NamespacedName, typedProvider); err != nil {
		if apierrors.IsNotFound(err) {
			// Object not found, return. Created objects are automatically garbage collected.
			// For additional cleanup logic use finalizers.
			return ctrl.Result{}, nil
		}
		// Error reading the object - requeue the request.
		return ctrl.Result{}, err
	}

	// Convert provider to a generic type so it can use shared functions
	genericProvider, err := r.toGenericProvider(typedProvider)

	// Initialize the patch helper
	patchHelper, err := patch.NewHelper(typedProvider, r.Client)
	if err != nil {
		return ctrl.Result{}, err
	}

	defer func() {
		if err := patchHelper.Patch(ctx, r.toTypedProvider(typedProvider)); err != nil {
			reterr = kerrors.NewAggregate([]error{reterr, err})
		}
	}()
	
	...
}
```

r.Object will gives us all the info to do conversion when necessary...

Reply:
thanks for idea!